### PR TITLE
[23] DOM Support for implicitly declared classes

### PR DESCRIPTION
**Pull request by subyssurendran666**
## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java [hunk 1]
```diff
@@ -1859,6 +1859,20 @@ public boolean visit(CompilationUnit node) {
...
+	@Override
+	public boolean visit(ImplicitTypeDeclaration node) {
+		if (!hasChildrenChanges(node)) {
+			return doVisitUnchangedChildren(node);
+		}
+		//javaDoc
+		int pos= rewriteJavadoc(node, ImplicitTypeDeclaration.JAVADOC_PROPERTY);
+
+		int startIndent= getIndent(node.getStartPosition()) + 1;
+		int startPos= getPosAfterLeftBrace(pos);
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java:1871]: Suby, I believe there is no left brace.  Shouldn't this be the beginning of first element in Body declarations ?

File: org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/rewrite/describing/ASTRewritingImplicitTypeDeclarationTest.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+import org.eclipse.jdt.core.dom.StringLiteral;
+import org.eclipse.jdt.core.dom.TagElement;
+import org.eclipse.jdt.core.dom.TextElement;
+import org.eclipse.jdt.core.dom.rewrite.ASTRewrite;
+import org.eclipse.jdt.core.dom.rewrite.ListRewrite;
+
+import junit.framework.Test;
+
+public class ASTRewritingImplicitTypeDeclarationTest extends ASTRewritingTest{
+
```
Comment (mpalat) [inline org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/rewrite/describing/ASTRewritingImplicitTypeDeclarationTest.java:42]: Can you please add a test that will remove the first method /function and add another method

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java (hunk 1 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java:1871]: For Implicit type declaration, the value is always zero. So can I hardcoded the value.

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/AST.java [hunk 3]
```diff
@@ -3190,6 +3190,23 @@ public TryStatement newTryStatement() {
 		return new TryStatement(this);
 	}
 
+	/**
+	 * Creates an unparented class declaration node owned by this AST.
+	 * The name of the class is an unspecified, but legal, name;
+	 * no modifiers; no doc comment; no superclass or superinterfaces;
+	 * and an empty class body.
+	 * <p>
+	 * To create an interface, use this method and then call
```
[the remark this answers is not in the document]
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/AST.java]: Done

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java (hunk 1 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.java]: Calculating the start position with `node.getStartPosition();`

File: org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/rewrite/describing/ASTRewritingImplicitTypeDeclarationTest.java (hunk 2 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/rewrite/describing/ASTRewritingImplicitTypeDeclarationTest.java:42]: Added one additional unit test based on the requirement.

**Review by mpalat — APPROVED** [review]

**Comment by stephan-herrmann** [comment]
@mpalat  @subyssurendran666  Changes on ASTConverter_23Test here and in #3252 (@jarthana fyi) cause issues when merging into BETA_JAVA24, notably because in the other branch this file has been renamed! (current status: compile errors after the merge - I'm currently fixing those).

From a systematic point of view, implicitly declared classes are still in preview. So shouldn't we make *all* changes regarding this feature in the beta branch??

**Comment by mpalat** [comment]
> From a systematic point of view, implicitly declared classes are still in preview. So shouldn't we make _all_ changes regarding this feature in the beta branch??

@stephan-herrmann Since this was applicable in master as well it was pushed to master. For example, issues in primitive types in pattern, we have pushed to master rather than Beta. However, I agree with you that from a systematic point of view, we can discuss and fix a method. I would think that:

1.  non-API changes of the preview features which can go into master if the issue owner wishes so and if it doesn't create any merge issues and failures and API changes (like this one) can go to BETA. 
2. Or as you mentioned, all changes can goto preview.

I agree that the second option is more uniform and zero-confusion and maybe easier to follow for everyone. If the general consensus is that this is better we can go alongwith this option.

@jarthana any comments from your side?

**Pull request merged**